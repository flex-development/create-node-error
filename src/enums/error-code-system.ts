/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
