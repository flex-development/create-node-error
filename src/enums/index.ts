/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
