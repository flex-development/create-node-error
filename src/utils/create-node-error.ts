/**
 * @file Utilities - createNodeError
 * @module errnode/utils/createNodeError
 * @see https://github.com/nodejs/node/blob/v19.3.0/lib/internal/errors.js#L367-L404
 */

import type { ErrorCode } from '#src/enums'
import type { NodeErrorConstructor } from '#src/interfaces'
import type { Args } from '#src/interfaces/node-error-constructor'
import kIsNodeError from '#src/internal/k-is-node-error'
import type { ErrnodeConstructor, Message, MessageFn } from '#src/types'
import { cast, define, isString, regexp } from '@flex-development/tutils'
import { format } from 'node-inspect-extracted'

/**
 * Creates a Node.js error constructor.
 *
 * Constructor arguments are passed to [`util.format`][1] if the error `message`
 * is a string, or `message` itself if it is a function. Message functions will
 * also be called with the new Node.js error instance as `this`.
 *
 * [1]: https://nodejs.org/api/util.html#utilformatformat-args
 *
 * @see {@linkcode NodeErrorConstructor}
 * @see https://nodejs.org/api/util.html#utilformatformat-args
 *
 * @template B - Error base class type
 * @template M - Error message function or `util.format` arguments
 * @template T - Error prototype
 *
 * @param {ErrorCode} code - Node.js error code
 * @param {B} Base - Error base class
 * @param {Message<M>} message - Error message string or function
 * @return {NodeErrorConstructor<T, M>} `NodeError` constructor
 */
const createNodeError = <
  B extends ErrnodeConstructor,
  M extends MessageFn | string | readonly unknown[],
  T extends B['prototype']
>(
  code: ErrorCode,
  Base: B,
  message: Message<M>
): NodeErrorConstructor<T, M> => {
  // @ts-expect-error ts(2322)
  return class NodeError extends Base {
    /**
     * Node.js error code.
     *
     * @public
     * @instance
     * @member {ErrorCode}
     */
    public code!: ErrorCode

    /**
     * Creates a new Node.js error.
     *
     * @param {any[]} args - Error message function arguments
     */
    constructor(...args: any[]) {
      super()

      /**
       * Instance properties are defined in order of precedence.
       *
       * Precedence is determined by latter property definitions using the value
       * of former property definitions to generate a value.
       *
       * Instance property dependencies:
       *
       * - `message`: `code`
       * - `toString`: `code`, `message`, `name`
       * - `stack`: `toString`
       */

      // add error symbol
      define(this, kIsNodeError, {
        configurable: true,
        enumerable: false,
        value: true,
        writable: false
      })

      // add error code
      define(this, 'code', {
        configurable: true,
        enumerable: true,
        value: code,
        writable: true
      })

      // add error message
      define(this, 'message', {
        configurable: true,
        enumerable: false,
        value: NodeError.#message(code, message, cast(args), this),
        writable: true
      })

      // redefine toString method
      define(this, 'toString', {
        configurable: true,
        enumerable: false,
        /**
         * Returns a string representation of the error.
         *
         * @this {import('#src/types').NodeError}
         *
         * @return {string} String representation of error
         */
        value(this: NodeError): string {
          return `${this.name} [${this.code}]: ${this.message}`
        },
        writable: true
      })

      // add stack trace
      define(this, 'stack', {
        configurable: true,
        enumerable: true,
        value: NodeError.#prepareStackTrace(this),
        writable: true
      })
    }

    /**
     * Creates an error message string.
     *
     * @see https://nodejs.org/api/util.html#utilformatformat-args
     *
     * @private
     * @static
     *
     * @param {ErrorCode} code - Node.js error code
     * @param {Message<M>} message - Error message function or format string
     * @param {Args<M>} args - Custom message function arguments if `message` is
     * a function, or `util.format` arguments if `message` is a string
     * @param {Error} self - Error object used as `this` argument
     * @return {string} Formatted error message
     * @throws {Error} If `args` length is invalid
     */
    static #message(
      code: ErrorCode,
      message: Message<M>,
      args: Args<M>,
      self: Error
    ): string {
      /**
       * Invalid {@linkcode args} length error message.
       *
       * @var {string} error
       */
      let error: string = `${code};` + ' '

      // check args.length against expected message function arguments length
      // default parameters do not count when message is a function
      if (isString(message)) {
        /**
         * {@linkcode format} specifiers regex.
         *
         * @const {RegExp} regex
         */
        const regex: RegExp = /%[Odfijos]/g

        /**
         * Expected message length.
         *
         * @var {number} length
         */
        let length: number = 0

        // get expected message length
        while (regex.exec(message) !== null) length++

        // throw if incorrect number of arguments are passed to util.format
        if (length !== args.length) {
          error += `The arguments length (${args.length}) provided to \`util.format\` does not match the required length (${length}).`
          throw new Error(error, { cause: args })
        }
      } else if (!(message.length <= args.length)) {
        error += `The arguments length (${args.length}) provided to \`message\` does not match the required length (${message.length}).`
        throw new Error(error, { cause: args })
      }

      return Reflect.apply(
        isString(message) ? format : message,
        isString(message) ? null : self,
        isString(message) ? [message, ...args] : args
      )
    }

    /**
     * Adds a formatted stack trace to the given `error`.
     *
     * Previously written stack traces are not preserved.
     *
     * @see https://v8.dev/docs/stack-trace-api#customizing-stack-traces
     *
     * @private
     * @static
     *
     * @param {Error} error - Error object to prepare stack trace for
     * @return {string} Formatted stack trace
     */
    static #prepareStackTrace(error: Error): string {
      // create stack property
      Base.captureStackTrace(error, NodeError.#prepareStackTrace)

      // format stack trace
      return error.stack!.replace(
        new RegExp(`^${error.name}: ${regexp(error.message)}`),
        error.toString()
      )
    }
  }
}

export default createNodeError
