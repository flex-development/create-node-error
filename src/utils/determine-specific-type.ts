/**
 * @file Utilities - determineSpecificType
 * @module errnode/utils/determineSpecificType
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L988-L1041
 */

import { inspect } from 'node-inspect-extracted'

/**
 * Determine the specific type of a value for type-mismatch errors.
 *
 * @param {unknown} value - Value to check
 * @return {string} Specific type of `value`
 */
function determineSpecificType(value: unknown): string {
  /**
   * Type of {@linkcode value}.
   *
   * @const {string} type
   */
  const type: string = typeof value

  /**
   * Specific type.
   *
   * @var {string} specificType
   */
  let specificType: string = ''

  switch (type) {
    case 'object':
    case 'undefined':
      specificType = value?.constructor && 'name' in value.constructor
        ? `an instance of ${value.constructor.name}`
        : inspect(value, { depth: -1 })
      break
    case 'function':
      specificType = `type ${type} ${(<(...args: any[]) => any>value).name}`
      break
    case 'string':
      // @ts-expect-error `value` is a string (18046).
      value.length > 28 && (value = `${(<string>value).slice(0, 25)}...`)
      specificType = `type ${type} (${inspect(value)})`
      break
    default:
      specificType = `type ${type} (${inspect(value)})`
      break
  }

  return specificType
}

export default determineSpecificType
