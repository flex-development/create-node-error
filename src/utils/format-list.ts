/**
 * @file Utilities - formatList
 * @module errnode/utils/formatList
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1043-L1060
 */

import { at, join } from '@flex-development/tutils'

/**
 * Create a list string in the form `'A and B'` or `'A, B, ..., and Z`.
 *
 * > {@linkcode Intl.ListFormat} cannot be used because it's not available in
 * > `--without-intl` builds.
 *
 * @example
 *  formatList([])
 *  // ''
 * @example
 *  formatList(['json'])
 *  // 'json'
 * @example
 *  formatList(['data', 'file'])
 *  // 'data and file'
 * @example
 *  formatList(['boolean', 'number', 'string'], 'or')
 *  // 'boolean, number, or string'
 *
 * @param {ReadonlyArray<string> | string} list
 *  List to format
 * @param {'and' | 'or' | null | undefined} [type='and']
 *  List transition type to insert before last element
 * @return {string}
 *  Formatted list
 */
function formatList(
  list: string | readonly string[],
  type?: 'and' | 'or' | null | undefined
): string {
  if (typeof list === 'string') return list

  if (!type) type = 'and'

  switch (list.length) {
    case 0:
      return ''
    case 1:
      return `${list[0]}`
    case 2:
      return `${list[0]} ${type} ${list[1]}`
    case 3:
      return `${list[0]}, ${list[1]}, ${type} ${list[2]}`
    default:
      return `${join(list.slice(0, -1), ', ')}, ${type} ${at(list, -1)}`
  }
}

export default formatList
