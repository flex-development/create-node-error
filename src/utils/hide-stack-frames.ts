/**
 * @file Utilities - hideStackFrames
 * @module errnode/utils/hideStackFrames
 */

import type { Fn } from '@flex-development/tutils'

/**
 * Remove unnecessary frames from Node.js core errors.
 *
 * @template {Fn} T
 *  Function to wrap
 *
 * @param {T} fn
 *  Function to wrap
 * @return {T}
 *  `fn`
 */
function hideStackFrames<T extends Fn>(fn: T): T {
  // rename functions that will be hidden.
  // this cuts off the stack trace at the outermost one.
  Object.defineProperty(fn, 'name', { value: '__node_internal_' + fn.name })
  return fn
}

export default hideStackFrames
