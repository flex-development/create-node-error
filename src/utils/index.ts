/**
 * @file Entry Point - Utilities
 * @module errnode/utils
 */

export { default as determineSpecificType } from './determine-specific-type'
export { default as formatList } from './format-list'
export { default as hideStackFrames } from './hide-stack-frames'
export { default as isNodeError } from './is-node-error'
