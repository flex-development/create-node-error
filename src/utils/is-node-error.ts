/**
 * @file Utilities - isNodeError
 * @module errnode/utils/isNodeError
 */

import kIsNodeError from '#k-is-node-error'
import type { NodeError } from '#src/interfaces'

/**
 * Check if `value` is a {@linkcode NodeError}.
 *
 * @param {unknown} value
 *  Value to check
 * @return {value is NodeError}
 *  `true` if `value` is `NodeError`, `false` otherwise
 */
function isNodeError(value: unknown): value is NodeError {
  return (
    typeof value === 'object' &&
    value !== null &&
    value instanceof Error &&
    kIsNodeError in value &&
    value[kIsNodeError] === true
  )
}

export default isNodeError
