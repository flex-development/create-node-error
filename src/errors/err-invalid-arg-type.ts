/**
 * @file Errors - ERR_INVALID_ARG_TYPE
 * @module errnode/errors/ERR_INVALID_ARG_TYPE
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1373-L1443
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'
import { determineSpecificType } from '#src/utils'
import formatList from '#src/utils/format-list'
import { DOT } from '@flex-development/tutils'
import { ok } from 'devlop'

/**
 * `ERR_INVALID_ARG_TYPE` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_invalid_arg_type
 *
 * @extends {NodeError<codes.ERR_INVALID_ARG_TYPE>}
 * @extends {TypeError}
 */
interface ErrInvalidArgType
  extends NodeError<codes.ERR_INVALID_ARG_TYPE>, TypeError {}

/**
 * `ERR_INVALID_ARG_TYPE` message arguments.
 */
type Args = [
  name: string,
  expected: string | readonly string[],
  actual: unknown
]

/**
 * `ERR_INVALID_ARG_TYPE` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrInvalidArgType}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrInvalidArgType,Args>}
 */
interface ErrInvalidArgTypeConstructor
  extends NodeErrorConstructor<ErrInvalidArgType, Args> {
  /**
   * Create a new `ERR_INVALID_ARG_TYPE` error.
   *
   * @see {@linkcode ErrInvalidArgType}
   *
   * @param {string} name
   *  Name of invalid argument or property
   * @param {ReadonlyArray<string> | string} expected
   *  Expected type(s)
   * @param {unknown} actual
   *  Value supplied by user
   * @return {ErrInvalidArgType}
   */
  new (
    name: string,
    expected: string | readonly string[],
    actual: unknown
  ): ErrInvalidArgType
}

/**
 * `ERR_INVALID_ARG_TYPE` model.
 *
 * Thrown when an argument of the wrong type is passed to a Node.js API.
 *
 * @see {@linkcode ErrInvalidArgTypeConstructor}
 *
 * @type {ErrInvalidArgTypeConstructor}
 *
 * @class
 */
const ERR_INVALID_ARG_TYPE: ErrInvalidArgTypeConstructor = E(
  codes.ERR_INVALID_ARG_TYPE,
  TypeError,
  /**
   * @param {string} name
   *  Name of invalid argument or property
   * @param {ReadonlyArray<string> | string} expected
   *  Expected type(s)
   * @param {unknown} actual
   *  Value supplied by user
   * @return {string}
   *  Error message
   */
  function message(
    name: string,
    expected: string | readonly string[],
    actual: unknown
  ): string {
    ok(typeof name === 'string', '\'name\' must be a string')

    if (typeof expected === 'string') expected = [expected]

    /**
     * Primitive value names.
     *
     * Sorted by a rough estimate on most frequently used entries.
     *
     * @const {Set<string>} kTypes
     */
    const kTypes: Set<string> = new Set([
      'string',
      'function',
      'number',
      'object',
      'Function',
      'Object',
      'boolean',
      'bigint',
      'symbol'
    ])

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = 'The '

    // stylize invalid argument name
    message += name.endsWith(' argument')
      ? name
      : `'${name}' ${name.includes(DOT) ? 'property' : 'argument'}`

    // continue building error message
    message += ' must be '

    /**
     * List of expected class instances.
     *
     * @const {string[]} instances
     */
    const instances: string[] = []

    /**
     * List of other expected types.
     *
     * @const {string[]} other
     */
    const other: string[] = []

    /**
     * List of expected primitive types.
     *
     * @const {string[]} types
     */
    const types: string[] = []

    // get expected types
    for (const value of expected) {
      ok(typeof value === 'string', 'values in `expected` must be strings')

      if (kTypes.has(value)) {
        types.push(value.toLowerCase())
      } else if (/^([A-Z][\da-z]*)+$/.exec(value)) {
        instances.push(value)
      } else {
        ok(value !== 'object', '"object" should be written as "Object"')
        other.push(value)
      }
    }

    // special case: handle `object` in case other instances are allowed to
    // outline the differences between each other
    if (instances.length) {
      /**
       * Position of `'object'` in {@linkcode types}.
       *
       * @const {number} position
       */
      const position: number = types.indexOf('object')

      // replace 'object' in types with 'Object' in instances
      if (position !== -1) {
        types.splice(position, 1)
        instances.push('Object')
      }
    }

    // add expected primitive types to error message
    if (types.length) {
      message += `${types.length > 1 ? 'one ' : ''}of type `
      message += formatList(types, 'or')
      if (instances.length || other.length) message += ' or '
    }

    // add expected class instances to error message
    if (instances.length) {
      message += 'an instance of ' + formatList(instances, 'or')
      if (other.length) message += ' or '
    }

    // add other expected types to error message
    if (other.length) {
      if (other.length > 1) {
        message += `one of ${formatList(other, 'or')}`
      } else {
        /* c8 ignore next */
        if (other[0]!.toLowerCase() !== other[0]) message += 'an '
        message += `${other[0]}`
      }
    }

    return `${message}. Received ${determineSpecificType(actual)}`
  }
)

export {
  ERR_INVALID_ARG_TYPE as default,
  type ErrInvalidArgType,
  type Args as ErrInvalidArgTypeArgs,
  type ErrInvalidArgTypeConstructor
}
