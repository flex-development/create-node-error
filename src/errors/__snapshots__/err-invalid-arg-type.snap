// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_INVALID_ARG_TYPE > should build message from parameters (0) 1`] = `"The json argument must be one of type boolean, number, or string or an instance of Array or Object or null. Received undefined"`;

exports[`unit:errors/ERR_INVALID_ARG_TYPE > should build message from parameters (1) 1`] = `"The 'opts.ip_addr' property must be of type string or an instance of IpAddress. Received null"`;

exports[`unit:errors/ERR_INVALID_ARG_TYPE > should build message from parameters (2) 1`] = `"The 'packageKey' argument must be one of "exports" or "imports". Received null"`;
